### PR TITLE
asset tutorial

### PR DESCRIPTION
**Pull request by dizcza**
Tutorial: https://elephant-docs-development.readthedocs.io/en/enh-asset_tutorial/tutorials/asset.html

### CODE REVIEW

File: doc/tutorials/asset.ipynb [hunk 1]
```diff
@@ -0,0 +1,933 @@
...
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.2. Analytical probability matrix\n",
+    "\n",
+    "The second step is to estimate the probability $P_{null}$ that non-zero entries in `imat` are occured by chance. The resultant `pmat` matrix is defined as the probability of having coincident spikes at bins `i` and `j` strictly lower than the observed overlap (`imat`), under the null hypothesis of independence of the input spike trains."
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:212]:
```suggestion
    "The second step is to estimate the probability $P_{null}$ that non-zero entries in `imat` occurred by chance. The resultant `pmat` matrix is defined as the probability of having strictly fewer coincident spikes at bins `i` and `j` strictly than the observed overlap (`imat`) under the null hypothesis of independence of the input spike trains."
```

File: doc/tutorials/asset.ipynb [hunk 2]
```diff
@@ -0,0 +1,933 @@
...
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.3. Joint probability matrix\n",
+    "\n",
+    "The third step is postprocessing of the analytical probability matrix `pmat`, obtained from the previous step. Centered at each (i,j) entry of `pmat` matrix, we apply a diagonal kernel with shape `filter_shape` and select the top `nr_largest` probabilities of (i,j) neighborhood (defined by `filter_shape`), and compute the significance of these `nr_largest` joint neighbor probabilities. The resultant `jmat` matrix is a \"dilated\" version of `imat`.\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:258]:
```suggestion
    "The third step is postprocessing of the analytical probability matrix `pmat` obtained from the previous step. Centered at each (i,j) entry of `pmat` matrix, we apply a diagonal kernel with shape `filter_shape`, select the top `nr_largest` probabilities of (i,j) neighborhood (defined by `filter_shape`), and compute the significance of these `nr_largest` joint neighbor probabilities. The resultant `jmat` matrix is a \"dilated\" version of `imat`.\n",
```

commata

File: doc/tutorials/asset.ipynb [hunk 3]
```diff
@@ -0,0 +1,933 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# ASSET: Analysis of Sequences of Synchronous Events in Massively Parallel Spike Trains\n",
+    "\n",
+    "The tutorial is based on the paper \\[[1](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004939)\\] and demonstrates a method of finding patterns of synchronous spike times (synfire chains) which cannot be revealed by measuring neuronal firing rates only.\n",
+    "\n",
+    "In this tutorial, we use 50 neurons per group in 10 successive groups of a synfire chain embedded in a balanced network simulation. For more information about the data and ASSET algorithm, refer to \\[[1](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004939)\\].\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:11]:
```suggestion
    "In this tutorial, we use 50 neurons per group in 10 successive groups of a synfire chain embedded in a balanced network simulation. For more information about the data and ASSET algorithm refer to \\[[1](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004939)\\].\n",
```

no comma here

File: doc/tutorials/asset.ipynb [hunk 4]
```diff
@@ -0,0 +1,933 @@
...
+    "\n",
+    "plt.rcParams['text.latex.preamble'] = r\"\\usepackage{subdepth}, \\usepackage{type1cm}\"\n",
+    "plt.rcParams['mathtext.fontset'] = 'cm'\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we download the data, packed in NixIO structure, from https://gin.g-node.org/INM-6/elephant-data"
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:87]:
```suggestion
    "First, we download the data, which is stored as a `nix` file, from https://gin.g-node.org/INM-6/elephant-data"
```

File: doc/tutorials/asset.ipynb [hunk 5]
```diff
@@ -0,0 +1,933 @@
...
+    "The true unknown permutation is stored in the `segment.annotations['spiketrain_ordering']`. **The goal is to recreate this permuatation from raw data with the statistical method ASSET.**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Applying ASSET analysis\n",
+    "\n",
+    "### 2.1. Intersection matrix\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:169]:
```suggestion
    "### 2.1. Intersection Matrix\n",
```

Title Case

File: doc/tutorials/asset.ipynb [hunk 6]
```diff
@@ -0,0 +1,933 @@
...
+      "text/plain": [
+       "<Figure size 864x864 with 2 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# 2.1.1) create ASSET analysis object\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:196]:
```suggestion
    "# 2.1.1) Create an ASSET Object\n",
```

Title Case, doubled "analysis"

File: doc/tutorials/asset.ipynb [hunk 7]
```diff
@@ -0,0 +1,933 @@
...
+    "# hint: try different binsizes, e.g. binsize=2.5, 3.5, 4.0 ms\n",
+    "asset_obj = asset.ASSET(spiketrains, bin_size=3*pq.ms)\n",
+    "\n",
+    "# 2.1.2) compute the intersection matrix\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:200]:
```suggestion
    "# 2.1.2) Intersection Matrix\n",
```

Title Case, consistent titles

File: doc/tutorials/asset.ipynb [hunk 8]
```diff
@@ -0,0 +1,933 @@
...
+    "imat = asset_obj.intersection_matrix()\n",
+    "plt.matshow(imat)\n",
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.2. Analytical probability matrix\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:210]:
```suggestion
    "### 2.2. Analytical Probability Matrix\n",
```

Title Case

File: doc/tutorials/asset.ipynb [hunk 9]
```diff
@@ -0,0 +1,933 @@
...
+    "pmat = asset_obj.probability_matrix_analytical(imat, kernel_width=50*pq.ms)\n",
+    "plt.matshow(pmat)\n",
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.3. Joint probability matrix\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:256]:
```suggestion
    "### 2.3. Joint Probability Matrix\n",
```

Title Case

File: doc/tutorials/asset.ipynb [hunk 10]
```diff
@@ -0,0 +1,933 @@
...
+   "source": [
+    "plt.matshow(jmat)\n",
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.4. Mask matrix\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:641]:
```suggestion
    "### 2.4. Mask Matrix\n",
```

Title Case

File: doc/tutorials/asset.ipynb [hunk 11]
```diff
@@ -0,0 +1,933 @@
...
+    "alpha = .99\n",
+    "mmat = asset_obj.mask_matrices([pmat, jmat], [alpha, alpha])\n",
+    "plt.matshow(mmat);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.5. Find clusters in the mask matrix\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:684]:
```suggestion
    "### 2.5. Find Clusters in the Mask Matrix\n",
```

Title Case

File: doc/tutorials/asset.ipynb [hunk 12]
```diff
@@ -0,0 +1,933 @@
...
+    "cmat = asset_obj.cluster_matrix_entries(mmat, max_distance=11, min_neighbors=10, stretch=5)\n",
+    "plt.matshow(cmat)\n",
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.6. Sequences of synchronous events\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:726]:
```suggestion
    "### 2.6. Sequences of Synchronous Events\n",
```

Title Case

File: doc/tutorials/asset.ipynb [hunk 13]
```diff
@@ -0,0 +1,933 @@
...
+    "plt.matshow(mmat);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.5. Find clusters in the mask matrix\n",
+    "\n",
+    "Each entry (i,j) of the mask matrix $M$ from the previous step is assigned to a cluster id. A cluster is constrained to have at least `min` number of associated elements with at most `eps` intra-distance (within the group). The cluster index, or the (i,j) entry of the resultant `cmat` matrix, is:\n",
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:686]:
```suggestion
    "Each entry (i,j) of the mask matrix $M$ from the previous step is assigned to a cluster id. A cluster is constrained to have at least `min_neighbors` number of associated elements with at most `eps` intra-distance (within the group). The cluster index, or the (i,j) entry of the resultant `cmat` matrix, is:\n",
    
    typo
```

File: doc/tutorials/asset.ipynb [hunk 14]
```diff
@@ -0,0 +1,933 @@
...
+    "plt.colorbar();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.6. Sequences of synchronous events\n",
+    "\n",
+    "Given the input spike trains, two arrays of bin edges from step 2.2 and clustered intersection matrix `cmat` from step 2.5, extract the sequences of synchronous events (synfire chains).\n"
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:728]:
```suggestion
    "Given the input spike trains, two arrays of bin edges from step 2.2 and the clustered intersection matrix `cmat` from step 2.5, extract the sequences of synchronous events (synfire chains).\n"
```

missing article

File: doc/tutorials/asset.ipynb [hunk 15]
```diff
@@ -0,0 +1,933 @@
...
...Qem6rV+Lh1n2IJV1u3YdY+WU/4C2sCQWkUKLz3KNddfH6nTngT5RLeEUml1WCUIC3+GHNpEyuc92GYwE3YSQUI6HgQtGuViT7KoZfhhgDQDq5rS5lrTYAiUjnqJlU1KexfAcjiID4MR3PAkEouFWqglAAAH+ivQAQr3TWH26ru9yWHsBLmI4HeJzVVRsAAADA6+jnApmDIBSQQokMMWYuNwAAAEpz23Rgu8qm2639XLemC/AypuMxHQ8OstPAWg3ptRr6G+0zvNDRAACklld/mAKIjRPTxZy6y3PpOqi8vm6i3w3AHZiOB6RBvFdPgu/j6gsAIB4EnADEKxn9T6vPSGa9RKAd8C5GQjESCg6yc5WnvKs7dt9PQwwAAOAMN/ez3LpwdkXpstPHDQayrPI62ue7+VgBmYa741kgCAUnlG787FxJKq+jYHc6n1s7IPAvOnn7kBcA4O+6MJ39rHimtbmhD2iVrpycnFDaoz1fWnnlhztJI1F+rrPcgiCUBYJQcEIsc9yl2BpGGly4BWVuH/ICAPxdF6Zz3+IZUeSGfI/WHw6mLdF00ydGoigrzmNNKMAlWOcJAABncYUbmcKr/Uon0m113sMbqLMzCyOhGAmFJKtoeHEiFWq8V32o2JFsXD3ah7yIH3UTnMJ5mXp+znM3j4Ryq0AgYPuu0PGIZeaB04ukI3HpKOdePbe8hOl4FghCISiZP4QSrdDKS0u057w6VBveRZnax295kcrAkN/yDu5B2Uo9P+d5pgShkl3/O5l2q7TGMrrKL2XTLwhC+RNBKAsEoRCUzErIqTnuEkEouAdlah+/5UUq98dveQf3oGylnp/zPJ2jNr1cJ6c6gGYVhMrJyXF0RBaSgyCUP7EmFJAi0a7CJKMDw3BiuIVX16BwAnkBAP6uC9PZ1yovX90+pbmiMpHM9Ef7rkGDBvm6bCJ+lIv0YSQUI6Gg1ETC7X5Hea+L9pxXb98LwH28fNUdCHL7j3MgGbw8cinZ3xdtTVY7MwaQftTZ/sR0PAsEoRDk9SBUsr4bAAhCAYA3EISy91m0NUB6MB0PyGAMNwXgRtRNAIBYWY2aKQ9tDeAuBKEAuatxijUtpRviaMNWGc4KwK5U1ofUTch0TEGBm7ipP1weu+kMBAIaNGgQ5xTgMkzHYzoeUiQZQ4GtPiORzwuiE5x85Ole5APgHM4v72OaEBL
```
Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:816]:
```suggestion
    "#### With the sequences of synchronous events `sses` we found a permutation of the spiketrains that reveals the synfire chain as in the ground truth ordering in the last Figure, shown below."
```

missing article

Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:809]:
```suggestion
    "plt.title('Reconstructed ordering of the neurons (y-axis) with synfire chains');"
```

consistent plot titles

Comment (Kleinjohann) [inline doc/tutorials/asset.ipynb:848]:
```suggestion
    "plt.title('True (unknown) ordering of the neurons (y-axis)')\n",
```

missing article

Comment (Kleinjohann) [reply doc/tutorials/asset.ipynb:816]: sorry, from here on the web interface for reviews mixed up / shifted the line numbers somehow

**Comment by dizcza** [comment]
Thanks a lot!
I also find a dozen really fast tqdm progress bars in `joint_probability_matrix` while we actually need only the last one, but it's the subject of asset improvements and we'll come to this (hopefully) shortly in the following PR.

**Comment by Kleinjohann** [comment]
> However, in order to be able to run it on mybinder I had to add `sklearn` and `nixio==1.5.0b3` to the environment. Do you know how we can change the environment creation?

This still seems to be the case, are you planning to fix this in another PR?

**Comment by dizcza** [comment]
This is something I did not notice, I forgot to run it on binder.
I'll address this before making a new release. Thanks.

**Pull request merged**